text/0041-extensible-block-body.md

```diff
@@ -0,0 +1,104 @@
+
+## Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+The hash of the block body would be redefined as the following:
+
+- hash of the body is the hash of `[hash of the deploy hashes list, hash_2]`
+- `hash_2` is the hash of `[hash of the transfer hashes list, hash_3]`
+- `hash_3` is the hash of `[hash of the proposer, hash_4]`
+- `hash_4` is the hash of an empty bytestring.
+
+This can be illustrated by the following diagram:
+
+```
+                   body_hash
+                   /      \
+ hash(deploy_hashes)      hash_2____________
+                         /                  \
+                   hash(transfer_hashes)     hash_3_______
+                                            /             \
+                                         hash(proposer)   hash("")
+```
+
+So, for example, to securely download only the transfer hashes stored in a block with a known body hash, we will additionaly require only the hash of the deploy hashes list and `hash_3`. This will allow us to calculate `hash_2` and `body_hash`, and to verify that the body hash matches the one we have.
+
+The block bodies would be stored the following way:
+- A `block_body_merkle` database would store a key-value map of `body_hash → vec![hash(deploy_hashes), hash(transfer_hashes), hash(proposer)]`. The type `Vec<Digest>` of the value will allow for storing blocks with more fields in the same database without the risk of causing problems with deserialization.
+- A `deploy_hashes` database would store the lists of deploy hashes keyed by `hash(deploy_hashes)`.
+- A `transfer_hashes` database would store the lists of transfer hashes keyed by `hash(transfer_hashes)`.
+- A `proposer` database would store the proposers keyed by `hash(proposer)`.
+
+The new way of hashing and new databases would start being used in a specific, chosen protocol version. The blocks in this protocol version and above would have their bodies hashed using the new way and be stored in the new databases. Older blocks would be kept in the old database.
+
+This is to ensure that we can still verify the integrity of storage after we transition to the new format. Changing the body hash of the old blocks would also change the hash of the block and hashes of all the later blocks, so it is not an option. So in order to be able to verify that the stored data still corresponds to the hashes from the blocks, we need to be able to calculate them the old way even after we start using the new way for new blocks.
+
+The code reading the blocks from storage would then have to be aware of the protocol version at which the transition happened, in order to read the block bodies from the correct databases regardless of whether it's a new or an old block.
```

Comment:
Do we need specialized storage for that? If we decide that the new abstraction is only in the way we compute the hash, we could store the block as it's done right now and calculates hashes "on the fly" – pull the block by hash, calculate the Merkle proof as necessary.

Reply:
But I think one of the points is to be able to have incomplete blocks (@xcthulhu correct me if I'm wrong) - in such a case, the same storage format will not work.

Reply:
What does "incomplete block" mean? Also, what's the use case? Can you give me an example? I work better with concrete examples.

Reply:
By "incomplete block" I mean a block in which you downloaded and stored only some of the fields, and not all of them. So, for example, with fast sync you might want to only download the list of deploys for the purpose of checking for replays, but ignore the other fields to conserve bandwidth (the other fields are currently just the proposer, so it's not a great benefit, but we're future-proofing here :wink:).

Reply:
My understanding of how fast-sync would use it is that it would download the header (it knows the correct block hash) and then request parts of that block's body (for example only transfers and hashes) – it can use the response (transfers, hashes and the merkle proof) to verify that it's indeed what it needs.

Reply:
> But I think one of the points is to be able to have incomplete blocks (@xcthulhu correct me if I'm wrong) - in such a case, the same storage format will not work.

To avoid replay attacks, we're going to need to download historical deploys, but that's just a fragment of a block.  We need a way to safely transmit fragments of a block.

Reply:
Fragments of a block's _body_ together with the Merkle proof but my proposal also achieves that without requiring any changes from the storage layer.

Reply:
> By "incomplete block" I mean a block in which you downloaded and stored only some of the fields, and not all of them. So, for example, with fast sync you might want to only download the list of deploys for the purpose of checking for replays, but ignore the other fields to conserve bandwidth (the other fields are currently just the proposer, so it's not a great benefit, but we're future-proofing here wink Or maybe I'm forgetting some other benefits).

I don't think you could consider this "incomplete block" to be a block at all. Would such node be useful for the network? Imagine a joining node trying to download the linear chain, it would query your node asking for block A only to find that you have "some" of block A and not all of it. 

Hashing the block body in the scheme you're proposing here makes sense for the extensibility purpose and for fast-sync (when it needs to download parts of the body to protect against the replay attack) but it does not respect the integrity of the data.